### PR TITLE
Use CLI11 for option parsing

### PR DESCRIPTION
**Pull request by mathiaswagner**
This is a proposal to switch option parsing for all tests to CLI11, https://github.com/CLIUtils/CLI11.

It is not fully functional yet, but a lot of things work. It is a lot less code and much more flexibility. Also just making sure to print the options actually applying to a test is possible.

For now I have only modified `staggered_dslash_test`.

The still not sorted or fully group `--help` looks like:

```
QUDA internal test
Usage: ./tests/staggered_dslash_test [OPTIONS]

Options:
  -h,--help                   Print this help message and exit
  --alternative-reliable BOOL use alternative reliable updates
  --anisotropy FLOAT          Temporal anisotropy factor (default 1.0)
  --ca-basis-type ENUM:value in {chebyshev->1,power->0} OR {1,0}
                              The basis to use for CA-CG (default power)
  --cheby-basis-eig-max FLOAT Conservative estimate of largest eigenvalue for Chebyshev basis CA-CG (default is to guess with power iterations)
  --cheby-basis-eig-min FLOAT Conservative estimate of smallest eigenvalue for Chebyshev basis CA-CG (default 0)
  --clover-coeff FLOAT        Clover coefficient (default 1.0)
  --compute-clover BOOL       Compute the clover field or use random numbers (default false)
  --compute-fat-long BOOL     Compute the fat/long field or use random numbers (default false)
  --contraction-type ENUM:value in {dr->1,open->0} OR {1,0}
                              Whether to leave spin elemental open, or use a gamma basis and contract on spin (default open)
  --dagger                    Set the dagger to 1 (default 0)
  --device INT:INT in [0 - 16]
                              Set the CUDA device to use (default 0, single GPU only)
  --dslash-type ENUM:value in {wilson->0,clover->1,twisted-mass->7,twisted-clover->8,staggered->5,asqtad->6,domain-wall->2,domain-wall-4d->3,mobius->4,laplace->9} OR {0,1,7,8,5,6,2,3,4,9}
                              Set the dslash type
  --epsilon FLOAT             Twisted-Mass flavor twist of Dirac operator (default 0.01)
  --epsilon-naik FLOAT        Epsilon factor on Naik term (default 0.0, suggested non-zero -0.1)
  --flavor ENUM:value in {singlet->1,deg-doublet->-2,nondeg-doublet->2,no->0} OR {1,-2,2,0}
                              Set the twisted mass flavor type (singlet (default), deg-doublet, nondeg-doublet)
  --gaussian-sigma FLOAT      Width of the Gaussian noise used for random gauge field contruction (default 0.2)
  --heatbath-beta FLOAT       Beta value used in heatbath test (default 6.2)
  --heatbath-coldstart BOOL   Whether to use a cold or hot start in heatbath test (default false)
  --heatbath-num-hb-per-step INT
                              Number of heatbath hits per heatbath step (default 5)
  --heatbath-num-or-per-step INT
                              Number of overrelaxation hits per heatbath step (default 5)
  --heatbath-num-steps INT    Number of measurement steps in heatbath test (default 10)
  --heatbath-warmup-steps INT Number of warmup steps in heatbath test (default 10)
  --inv-type ENUM:value in {cg->0,bicgstab->1,gcr->2,pcg->7,mpcg->8,mpbicgstab->4,mr->3,sd->5,eigcg->9,inc-eigcg->10,gmresdr->11,gmresdr-proj->12,gmresdr-sh->13,fgmresdr->14,mg->15,bicgstab-l->16,cgne->17,cgnr->18,cg3->19,cg3ne->20,cg3nr->21,ca-cg->22,ca-cgne->23,ca-cgnr->24,ca-gcr->25} OR {0,1,2,7,8,4,3,5,9,10,11,12,13,14,15,16,17,18,19,20,21,22,23,24,25}
                              The type of solver to use (default cg)
  --kappa FLOAT               Kappa of Dirac operator (default 0.12195122... [equiv to mass])
  --laplace3D INT             Restrict laplace operator to omit the t dimension (n=3), or include all dims (n=4) (default 4)
  --load-gauge TEXT           Load gauge field " file " for the test (requires QIO)
  --Lsdim INT                 Set Ls dimension size(default 16)
  --mass FLOAT                Mass of Dirac operator (default 0.1)
  --mass-normalization ENUM:value in {kappa->0,mass->1,asym-mass->2} OR {0,1,2}
                              Mass normalization (kappa (default) / mass / asym-mass)
  --matpc ENUM:value in {even-even->0,odd-odd->1,even-even-asym->2,odd-odd-asym->3} OR {0,1,2,3}
                              Matrix preconditioning type (even-even, odd-odd, even-even-asym, odd-odd-asym)
  --msrc INT                  Used for testing non-square block blas routines where nsrc defines the other dimension
  --mu FLOAT                  Twisted-Mass chiral twist of Dirac operator (default 0.1)
  --multishift INT            Whether to do a multi-shift solver test or not (default false)
  --ngcrkrylov INT            The number of inner iterations to use for GCR, BiCGstab-l, CA-CG (default 10)
  --niter INT                 The number of iterations to perform (default 10)
  --nsrc INT                  How many spinors to apply the dslash to simultaneusly (experimental for staggered only)
  --pipeline INT              The pipeline length for fused operations in GCR, BiCGstab-l (default 0, no pipelining)
  --prec ENUM:value in {double->8,half->2,single->4} OR {8,2,4}
                              Precision in GPU
  --prec-precondition ENUM:value in {double->8,half->2,single->4} OR {8,2,4}
                              Preconditioner precision in GPU
  --prec-refine ENUM:value in {double->8,half->2,single->4} OR {8,2,4}
                              Sloppy precision for refinement in GPU
  --prec-ritz ENUM:value in {double->8,half->2,single->4} OR {8,2,4}
                              Eigenvector precision in GPU
  --prec-sloppy ENUM:value in {double->8,half->2,single->4} OR {8,2,4}
                              Sloppy precision in GPU
  --prec-null ENUM:value in {double->8,half->2,single->4} OR {8,2,4}
                              Precison TODO
  --precon-type ENUM:value in {cg->0,bicgstab->1,gcr->2,pcg->7,mpcg->8,mpbicgstab->4,mr->3,sd->5,eigcg->9,inc-eigcg->10,gmresdr->11,gmresdr-proj->12,gmresdr-sh->13,fgmresdr->14,mg->15,bicgstab-l->16,cgne->17,cgnr->18,cg3->19,cg3ne->20,cg3nr->21,ca-cg->22,ca-cgne->23,ca-cgnr->24,ca-gcr->25} OR {0,1,2,7,8,4,3,5,9,10,11,12,13,14,15,16,17,18,19,20,21,22,23,24,25}
                              The type of solver to use (default none (=unspecified)).
  --rank-order INT            Set the [t][z][y][x] rank order as either column major (t fastest, default) or row major (x fastest)
  --recon ENUM:value in {18->18,13->13,12->12,9->9,8->8} OR {18,13,12,9,8}
                              Link reconstruction type
  --recon-precondition ENUM:value in {18->18,13->13,12->12,9->9,8->8} OR {18,13,12,9,8}
                              Preconditioner link reconstruction type
  --recon-sloppy ENUM:value in {18->18,13->13,12->12,9->9,8->8} OR {18,13,12,9,8}
                              Sloppy link reconstruction type
  --reliable-delta FLOAT      Set reliable update delta factor
  --save-gauge TEXT           Save gauge field " file " for the test (requires QIO, heatbath test only)
  --solution-pipeline INT     The pipeline length for fused solution accumulation (default 0, no pipelining)
  --solution-type ENUM:value in {mat->0,mat-dag-mat->1,mat-pc->2,mat-pc-dag->3,mat-pc-dag-mat-pc->4} OR {0,1,2,3,4}
                              The solution we desire (mat (default), mat-dag-mat, mat-pc, mat-pc-dag-mat-pc (default for multi-shift))
  --solve-type ENUM:value in {direct->0,direct-pc->2,normop->1,normop-pc->3,normerr->4,normerr-pc->5} OR {0,2,1,3,4,5}
                              The type of solve to do (direct, direct-pc, normop, normop-pc, normerr, normerr-pc)
  --solver-ext-lib-type ENUM:value in {eigen->1,magma->2} OR {1,2}
                              Set external library for the solvers  (default Eigen library)
  --tadpole-coeff FLOAT       Tadpole coefficient for HISQ fermions (default 1.0, recommended [Plaq]^1/4)
  --test INT                  Test method (different for each test)
  --tol FLOAT                 Set L2 residual tolerance
  --tolhq FLOAT               Set heavy-quark residual tolerance
  --unit-gauge BOOL           Generate a unit valued gauge field in the tests. If false, a random gauge is generated (default false)
  --verbosity ENUM:value in {silent->0,summarize->1,verbose->2,debug->3} OR {0,1,2,3}
                              The the verbosity on the top level of QUDA( default summarize)
  --verify BOOL               Verify the GPU results using CPU results (default true)
  --xdim INT:INT in [1 - 512] Set X dimension size(default 24)
  --ydim INT:INT in [1 - 512] Set X dimension size(default 24)
  --zdim INT:INT in [1 - 512] Set X dimension size(default 24)
  --tdim INT:INT in [1 - 512] Set T dimension size(default 24)
  --sdim INT:INT in [1 - 512] Set space dimension(X/Y/Z) size
[Option Group: Eigensolver]
  Options controlling eigensolver
  Options:
    --eig-amax FLOAT:POSITIVE   The maximum in the polynomial acceleration
    --eig-amin FLOAT:POSITIVE   The minimum in the polynomial acceleration
    --eig-ARPACK-logfile TEXT   ...
    --eig-QUDA-logfile TEXT     The filename storing the stdout from the QUDA eigensolver
    --eig-arpack-check BOOL     Cross check the device data against ARPACK (requires ARPACK, default false)
    --eig-check-interval INT    Perform a convergence check every nth restart/iteration in the IRAM,IRLM/lanczos,arnoldi eigensolver
    --eig-compute-svd BOOL      Solve the MdagM problem, use to compute SVD of M (default false)
    --eig-max-restarts INT      Perform n iterations of the restart in the eigensolver
    --eig-nConv INT             The number of converged eigenvalues requested
    --eig-nEv INT               The size of eigenvector search space in the eigensolver
    --eig-nKr INT               The size of the Krylov subspace to use in the eigensolver
    --eig-poly-deg INT          ...
    --eig-require-convergence BOOL
                                If true, the solver will error out if convergence is not attained. If false, a warning will be given (default true)
    --eig-save-vec TEXT         Save eigenvectors to <file> (requires QIO)
    --eig-load-vec TEXT:FILE    Load eigenvectors to <file> (requires QIO)
    --eig-spectrum ENUM:value in {LI->5,LM->3,LR->1,SI->4,SM->2,SR->0} OR {5,3,1,4,2,0}
                                The spectrum part to be calulated. S=smallest L=largest R=real M=modulus I=imaginary
    --eig-tol FLOAT             The tolerance to use in the eigensolver
    --eig-type ENUM:value in {iram->2,irlm->1,trlm->0} OR {2,1,0}
                                The type of eigensolver to use (default trlm)
    --eig-use-dagger BOOL       Solve the Mdag  problem instead of M (MMdag if eig-use-normop == true) (default false)
    --eig-use-normop BOOL       Solve the MdagM problem instead of M (MMdag if eig-use-dagger == true) (default false)
    --eig-use-poly-acc BOOL     Use Chebyshev polynomial acceleration in the eigensolver
[Option Group: Deflation]
  Options controlling deflation
  Options:
    --df-deflation-grid INT:POSITIVE
                                Set maximum number of cycles needed to compute eigenvectors(default 1)
    --df-eigcg-max-restarts INT:POSITIVE
                                Set how many iterative refinement cycles will be solved with eigCG within a single physical right hand site solve (default 4)
    --df-ext-lib-type ENUM      Set external library for the deflation methods  (default Eigen library)
    --df-location-ritz ENUM     Set memory location for the ritz vectors  (default cuda memory location)
    --df-max-restart-num INT    Set maximum number of the initCG restarts in the deflation stage (default 3)
    --df-max-search-dim INT     Set the size of eigenvector search space (default 64)
    --df-mem-type-ritz ENUM     Set memory type for the ritz vectors  (default device memory type)
    --df-nev INT                Set number of eigenvectors computed within a single solve cycle (default 8)
    --df-tol-eigenval FLOAT     Set maximum eigenvalue residual norm (default 1e-1)
    --df-tol-inc FLOAT          Set tolerance for the subsequent restarts in the initCG solver  (default 1e-2)
    --df-tol-restart FLOAT      Set tolerance for the first restart in the initCG solver(default 5e-5)
```

Couple of things remaining to be done, but wanted to put it out first:

* [x] add multi-grid options
* [x] add transformers for all used enums
* [x] group options 
* [x] modify all tests
* [x] Use `CLI::TransformPairs<T>` or `std::map<std::string,T>` for transformations (affects ordering in help)
* [x] remove old C-style option parsing

Later we could modify
* [ ] only add options relevant for the tests
* [ ] sort options
* [ ] ensure required options etc

### CODE REVIEW

**Comment by weinbe2** [comment]
Broadly, I like the direction this is going in!

Posing a suggestion: in cases where there are `ENUM`s, can we keep the standing behavior that the user has to specify the text string? Concrete example:

`--verbosity ENUM:value in {silent->0,summarize->1,verbose->2,debug->3} OR {0,1,2,3}`

Get rid of support for passing in 0, 1, 2, or 3. My justification is it'll make passing around reference commands simpler: I know what "silent" means, I don't want to have to figure out what "0" means. (Maybe verbosity is a bad example of this, but for things like dslash types, it could make a life-or-death sanity difference.)

**Comment by cpviolator** [comment]
I agree with @weinbe2 on this. Forcing uses (experienced and new) to use the more descriptive options will completely cut out any possible confusion over command line params.

Say for example one user gets slick and uses a series of digits as options. The next user who inherits the scripts is utterly confused by the numbers. On the other hand, if each option itself a description of what it does, it becomes self explanatory.

**Comment by mathiaswagner** [comment]
I need to look into being able to actually disallow passing integers. I don't want to go to custom or hack CLI11 to allow for easy updating.

Would modifying the help message be good enough? If that is easier ...

**Comment by mathiaswagner** [comment]
How about 
```
QUDA internal test
Usage: ./tests/staggered_dslash_test [OPTIONS]

Options:
  -h,--help                   Print this help message and exit
  --alternative-reliable BOOL use alternative reliable updates
  --anisotropy FLOAT          Temporal anisotropy factor (default 1.0)
  --ca-basis-type ENUM:value in {chebyshev,power}
                              The basis to use for CA-CG (default power)
  --cheby-basis-eig-max FLOAT Conservative estimate of largest eigenvalue for Chebyshev basis CA-CG (default is to guess with power iterations)
  --cheby-basis-eig-min FLOAT Conservative estimate of smallest eigenvalue for Chebyshev basis CA-CG (default 0)
  --clover-coeff FLOAT        Clover coefficient (default 1.0)
  --compute-clover BOOL       Compute the clover field or use random numbers (default false)
  --compute-fat-long BOOL     Compute the fat/long field or use random numbers (default false)
  --contraction-type ENUM:value in {dr,open}
                              Whether to leave spin elemental open, or use a gamma basis and contract on spin (default open)
  --dagger                    Set the dagger to 1 (default 0)
  --device INT:INT in [0 - 16]
                              Set the CUDA device to use (default 0, single GPU only)
  --dslash-type ENUM:value in {wilson,clover,twisted-mass,twisted-clover,staggered,asqtad,domain-wall,domain-wall-4d,mobius,laplace}
                              Set the dslash type
  --epsilon FLOAT             Twisted-Mass flavor twist of Dirac operator (default 0.01)
  --epsilon-naik FLOAT        Epsilon factor on Naik term (default 0.0, suggested non-zero -0.1)
  --flavor ENUM:value in {singlet,deg-doublet,nondeg-doublet,no}
                              Set the twisted mass flavor type (singlet (default), deg-doublet, nondeg-doublet)
  --gaussian-sigma FLOAT      Width of the Gaussian noise used for random gauge field contruction (default 0.2)
  --heatbath-beta FLOAT       Beta value used in heatbath test (default 6.2)
  --heatbath-coldstart BOOL   Whether to use a cold or hot start in heatbath test (default false)
  --heatbath-num-hb-per-step INT
                              Number of heatbath hits per heatbath step (default 5)
  --heatbath-num-or-per-step INT
                              Number of overrelaxation hits per heatbath step (default 5)
  --heatbath-num-steps INT    Number of measurement steps in heatbath test (default 10)
  --heatbath-warmup-steps INT Number of warmup steps in heatbath test (default 10)
  --inv-type ENUM:value in {cg,bicgstab,gcr,pcg,mpcg,mpbicgstab,mr,sd,eigcg,inc-eigcg,gmresdr,gmresdr-proj,gmresdr-sh,fgmresdr,mg,bicgstab-l,cgne,cgnr,cg3,cg3ne,cg3nr,ca-cg,ca-cgne,ca-cgnr,ca-gcr}
                              The type of solver to use (default cg)
  --kappa FLOAT               Kappa of Dirac operator (default 0.12195122... [equiv to mass])
  --laplace3D INT             Restrict laplace operator to omit the t dimension (n=3), or include all dims (n=4) (default 4)
  --load-gauge TEXT           Load gauge field " file " for the test (requires QIO)
  --Lsdim INT                 Set Ls dimension size(default 16)
  --mass FLOAT                Mass of Dirac operator (default 0.1)
  --mass-normalization ENUM:value in {kappa,mass,asym-mass}
                              Mass normalization (kappa (default) / mass / asym-mass)
  --matpc ENUM:value in {even-even,odd-odd,even-even-asym,odd-odd-asym}
                              Matrix preconditioning type (even-even, odd-odd, even-even-asym, odd-odd-asym)
  --msrc INT                  Used for testing non-square block blas routines where nsrc defines the other dimension
  --mu FLOAT                  Twisted-Mass chiral twist of Dirac operator (default 0.1)
  --multishift INT            Whether to do a multi-shift solver test or not (default false)
  --ngcrkrylov INT            The number of inner iterations to use for GCR, BiCGstab-l, CA-CG (default 10)
  --niter INT                 The number of iterations to perform (default 10)
  --nsrc INT                  How many spinors to apply the dslash to simultaneusly (experimental for staggered only)
  --pipeline INT              The pipeline length for fused operations in GCR, BiCGstab-l (default 0, no pipelining)
  --prec ENUM:value in {double,half,single}
                              Precision in GPU
  --prec-precondition ENUM:value in {double,half,single}
                              Preconditioner precision in GPU
  --prec-refine ENUM:value in {double,half,single}
                              Sloppy precision for refinement in GPU
  --prec-ritz ENUM:value in {double,half,single}
                              Eigenvector precision in GPU
  --prec-sloppy ENUM:value in {double,half,single}
                              Sloppy precision in GPU
  --prec-null ENUM:value in {double,half,single}
                              Precison TODO
  --precon-type ENUM:value in {cg,bicgstab,gcr,pcg,mpcg,mpbicgstab,mr,sd,eigcg,inc-eigcg,gmresdr,gmresdr-proj,gmresdr-sh,fgmresdr,mg,bicgstab-l,cgne,cgnr,cg3,cg3ne,cg3nr,ca-cg,ca-cgne,ca-cgnr,ca-gcr}
                              The type of solver to use (default none (=unspecified)).
  --rank-order INT            Set the [t][z][y][x] rank order as either column major (t fastest, default) or row major (x fastest)
  --recon ENUM:value in {18,13,12,9,8}
                              Link reconstruction type
  --recon-precondition ENUM:value in {18,13,12,9,8}
                              Preconditioner link reconstruction type
  --recon-sloppy ENUM:value in {18,13,12,9,8}
                              Sloppy link reconstruction type
  --reliable-delta FLOAT      Set reliable update delta factor
  --save-gauge TEXT           Save gauge field " file " for the test (requires QIO, heatbath test only)
  --solution-pipeline INT     The pipeline length for fused solution accumulation (default 0, no pipelining)
  --solution-type ENUM:value in {mat,mat-dag-mat,mat-pc,mat-pc-dag,mat-pc-dag-mat-pc}
                              The solution we desire (mat (default), mat-dag-mat, mat-pc, mat-pc-dag-mat-pc (default for multi-shift))
  --solve-type ENUM:value in {direct,direct-pc,normop,normop-pc,normerr,normerr-pc}
                              The type of solve to do (direct, direct-pc, normop, normop-pc, normerr, normerr-pc)
  --solver-ext-lib-type ENUM:value in {eigen,magma}
                              Set external library for the solvers  (default Eigen library)
  --tadpole-coeff FLOAT       Tadpole coefficient for HISQ fermions (default 1.0, recommended [Plaq]^1/4)
  --test INT                  Test method (different for each test)
  --tol FLOAT                 Set L2 residual tolerance
  --tolhq FLOAT               Set heavy-quark residual tolerance
  --unit-gauge BOOL           Generate a unit valued gauge field in the tests. If false, a random gauge is generated (default false)
  --verbosity ENUM:value in {silent,summarize,verbose,debug}
                              The the verbosity on the top level of QUDA( default summarize)
  --verify BOOL               Verify the GPU results using CPU results (default true)
  --xdim INT:INT in [1 - 512] Set X dimension size(default 24)
  --ydim INT:INT in [1 - 512] Set X dimension size(default 24)
  --zdim INT:INT in [1 - 512] Set X dimension size(default 24)
  --tdim INT:INT in [1 - 512] Set T dimension size(default 24)
  --sdim INT:INT in [1 - 512] Set space dimension(X/Y/Z) size
```

Not yet committed and might break sometimes. We could also allow the integer for some selected options.

**Comment by cpviolator** [comment]
IMO, this is slightly worse. If a user comes across a script that uses just integer digits for operators/verbosity etc, the help output doesn't give any help on what those integers do. The user must then trawl through source code to find out what they do.

If forcing descriptive arguments is difficult or undesirable, then I guess the the first pass you did is good as one can get as the help output gives both possible methods of input. I might suggest though that when we put command line examples on the Wiki, we always use the most descriptive option available rather than integers, and try to encourage users to do the same.

**Comment by mathiaswagner** [comment]
The last version I have and posted doesn’t accept integers. Or to be more precise: We can choose between string only or the old string or int behavior on a per-option basis.

**Comment by mathiaswagner** [comment]
@maddyscientist thoughts on that?

**Comment by maddyscientist** [comment]
This does look like a nice library and would be a good addition to QUDA.  I think I'm all in favour.

If we adopt this, does it in anyway limit our hand if we subsequently want to add some kind of parameter serialization support to QUDA?

**Comment by mathiaswagner** [comment]
Like reading them from a file?

**Comment by maddyscientist** [comment]
Yes

**Comment by mathiaswagner** [comment]
It already comes with https://github.com/CLIUtils/CLI11#configuration-file

**Comment by maddyscientist** [comment]
Similar comment here as with the stack trace printing.  This pull will add >6000 lines of external library code to QUDA.  Can we consider installing it using cmake?

One option could be to fork the repositories of these external libraries into the lattice org, so we always have these codes just in case the repository disappears.  Then we could install them automatically with cmake regardless.

**Comment by mathiaswagner** [comment]
Conpared to stack traces this is core functionality with some added code for things like the mg options which take a level + value. I’m not entirely convinced about making that a download at cmake time. Seems brittle. On the other hand we do that for eigen and that’s core as well. 
We could maintain our own fork and do submodules. Just afraid of things not working and the required user support. 

Need to think about that more.

**Comment by mathiaswagner** [comment]
This should be mostly feature complete. As always, there is room for improvement. (Better help messages, only adding options that apply, more fancy checks at the parameter level).
But that can be left for later iterations on this. It should now provide the functionality as before.

**Comment by mathiaswagner** [comment]
ping - any thoughts one this?

**Comment by weinbe2** [comment]
I can play with this this week. I want to get the staggered MG pull opened, but once that's done I can take a separate branch and try merging this in and seeing how hard it is for me to add this type of command arg offloading into it.

**Comment by mathiaswagner** [comment]
@weinbe2 @maddyscientist @cpviolator thoughts?

**Comment by cpviolator** [comment]
Let me spin this up and try out some of the test MG commands in the wiki, see how it goes.

**Comment by cpviolator** [comment]
I think this would be a lot easier to use and certainly makes the test scripts look a lot cleaner!

I was looking through the CLI11 documentation and they suggested the `rang` library, [here](https://github.com/CLIUtils/CLI11#other-libraries). I think it would be a nice addition to the help output as it can highlight the option groups, give error messages a distinct colour, etc. It's also a header only file. @mathiaswagner where in the code is the download for CLI11.hpp?

**Comment by mathiaswagner** [comment]
currently the header is copied in the rep, not downloaded. It is also slightly modified to be able to get a bit closer to the legacy implementation.

Rang looks promising but I would do that in a separate pull.

You can check what we do for backward-cpp (QUDA_BACKWARDS) for downloading in CMakeLists.txt.

File: lib/unitarize_links_quda.cu [hunk 1]
```diff
@@ -305,7 +305,7 @@ namespace{
     int num_failures = 0;
     Matrix<complex<double>,3> inlink, outlink;
       
-    for (int i=0; i<infield.Volume(); ++i){
+    for (unsigned int i=0; i<infield.Volume(); ++i){
```
Comment (weinbe2) [inline lib/unitarize_links_quda.cu]: I feel like we've addressed this with `size_t` in other places. @maddyscientist what's the good word on this?

Comment (maddyscientist) [reply lib/unitarize_links_quda.cu]: No need for more than 32 bits here unless we care about 256^4 volume per process.  Certainly no harm in being future proof....

Comment (mathiaswagner) [reply lib/unitarize_links_quda.cu]: I think the `unsigned` came in to silence a warning.

**Comment by weinbe2** [comment]
To help with tracking down MG-related bugs, here's a command that hits a large number of bases (of note missing eigensolver commands for now, but I believe those all get covered by other datatypes here already).

```
ARGSTR="--dim 16 16 16 16 --prec double --prec-sloppy single --prec-precondition half --prec-null half --mg-smoother-halo-prec single --dslash-type wilson --solve-type direct --verbosity verbose --nsrc 1 "
ARGSTR=$ARGSTR"--mass 0.0 "
ARGSTR=$ARGSTR"--mg-levels 4 --mg-coarse-solve-type 0 direct "
ARGSTR=$ARGSTR"--mg-setup-inv 1 cgnr --mg-setup-maxiter 1 1000 --mg-setup-tol 1 1e-5 "
ARGSTR=$ARGSTR"--mg-block-size 0 2 2 2 2 --mg-nvec 0 24 --mg-coarse-solve-type 1 direct-pc --mg-smoother-solve-type 0 direct    --mg-smoother 0 ca-gcr --mg-nu-pre 0 0 --mg-nu-post 0 8 --mg-smoother-tol 0 1e-10 --mg-coarse-solver-tol 1 5e-2 --mg-coarse-solver-maxiter 1 16   --mg-coarse-solver 1 gcr "
ARGSTR=$ARGSTR"--mg-block-size 1 2 2 2 2 --mg-nvec 1 24 --mg-n-block-ortho 1 2 --mg-coarse-solve-type 2 direct-pc --mg-smoother-solve-type 1 direct-pc --mg-smoother 1 ca-gcr --mg-nu-pre 1 0 --mg-nu-post 1 8 --mg-smoother-tol 1 1e-10 --mg-coarse-solver-tol 2 0.25 --mg-coarse-solver-maxiter 2 16   --mg-coarse-solver 2 gcr "
ARGSTR=$ARGSTR"--mg-block-size 2 2 2 2 2 --mg-nvec 2 32 --mg-n-block-ortho 2 2 --mg-coarse-solve-type 3 direct-pc --mg-smoother-solve-type 2 direct-pc --mg-smoother 2 ca-gcr --mg-nu-pre 2 0 --mg-nu-post 2 8 --mg-smoother-tol 2 1e-10 --mg-coarse-solver-tol 3 0.25 --mg-coarse-solver-maxiter 3 1024   --mg-coarse-solver 3 cgne "
./multigrid_invert_test $ARGSTR
```

**Pull request merged**